File: backend/app/core/entitlements.py
"""Plan-based feature entitlements + usage limits.

Single source of truth for what each subscription plan unlocks. The frontend
mirrors this in ``lib/entitlements.ts`` for UI gating; the backend enforces it
(403) so it can't be bypassed.
"""
from __future__ import annotations

from app.models.enums import PlanTier

PLAN_RANK: dict[PlanTier, int] = {
    PlanTier.starter: 1,
    PlanTier.growth: 2,
    PlanTier.scale: 3,
}

# Feature key -> the minimum plan that unlocks it. Unlisted features are open
# to every plan.
FEATURE_MIN_PLAN: dict[str, PlanTier] = {
    "forecasting": PlanTier.growth,
    "shipping": PlanTier.growth,
    "payments": PlanTier.growth,
    "realtime": PlanTier.scale,
    "integration:google_ads": PlanTier.growth,
    "integration:razorpay": PlanTier.growth,
    "integration:cashfree": PlanTier.growth,
    "integration:shiprocket": PlanTier.growth,
}

STORE_LIMIT: dict[PlanTier, int] = {
    PlanTier.starter: 1,
    PlanTier.growth: 3,
    PlanTier.scale: 9999,
}

TEAM_LIMIT: dict[PlanTier, int] = {
    PlanTier.starter: 2,
    PlanTier.growth: 5,
    PlanTier.scale: 9999,
}


def plan_allows(plan: PlanTier, feature: str) -> bool:
    minimum = FEATURE_MIN_PLAN.get(feature)
    if minimum is None:
        return True
    return PLAN_RANK.get(plan, 0) >= PLAN_RANK[minimum]


def team_limit(plan: PlanTier) -> int:
    return TEAM_LIMIT.get(plan, 2)

File: Frontend/src/lib/entitlements.ts
// ============================================================
// Plan entitlements (mirror of backend app/core/entitlements.py).
// UI gating only — the backend enforces the same rules with 403.
// ============================================================

export type Plan = "starter" | "growth" | "scale";

const RANK: Record<Plan, number> = { starter: 1, growth: 2, scale: 3 };

// Feature key -> minimum plan. Unlisted features are open to all plans.
const FEATURE_MIN: Record<string, Plan> = {
  forecasting: "growth",
  shipping: "growth",
  payments: "growth",
  realtime: "scale",
  "integration:google_ads": "growth",
  "integration:razorpay": "growth",
  "integration:cashfree": "growth",
  "integration:shiprocket": "growth",
};

export function planAllows(plan: string | undefined | null, feature: string): boolean {
  const min = FEATURE_MIN[feature];
  if (!min) return true;
  const p = (plan as Plan) || "starter";
  return (RANK[p] ?? 0) >= RANK[min];
}

export function requiredPlan(feature: string): Plan {
  return FEATURE_MIN[feature] ?? "starter";
}

// Sidebar href -> feature key (for nav lock indicators).
export const NAV_FEATURE: Record<string, string> = {
  "/dashboard/forecasting": "forecasting",
  "/dashboard/shipping": "shipping",
  "/dashboard/payments": "payments",
};